pkg/querier/dispatcher.go

```diff
@@ -0,0 +1,380 @@
+	stream frontendv2pb.QueryResultStream
+	logger log.Logger
+}
+
+func (w *queryResponseWriter) Write(ctx context.Context, m querierpb.EvaluateQueryResponse) error {
+	return w.stream.Write(ctx, &frontendv2pb.QueryResultStreamRequest{
+		Data: &frontendv2pb.QueryResultStreamRequest_EvaluateQueryResponse{
+			EvaluateQueryResponse: &m,
+		},
+	})
+}
+
+func (w *queryResponseWriter) WriteError(ctx context.Context, typ mimirpb.QueryErrorType, msg string) {
+	writeErrorToStream(ctx, w.stream, typ, msg, w.logger)
+}
+
+type evaluationObserver struct {
+	w         *queryResponseWriter
+	nodeIndex int64 // FIXME: remove this once Evaluator supports multiple nodes and passes the node index to the methods below
+
+	originalExpression string
+}
+
+func (o *evaluationObserver) SeriesMetadataEvaluated(ctx context.Context, evaluator *streamingpromql.Evaluator, series []types.SeriesMetadata) error {
+	defer types.SeriesMetadataSlicePool.Put(&series, evaluator.MemoryConsumptionTracker)
+
+	protoSeries := make([]querierpb.SeriesMetadata, 0, len(series))
+
+	for _, s := range series {
+		protoSeries = append(protoSeries, querierpb.SeriesMetadata{
+			Labels: mimirpb.FromLabelsToLabelAdapters(s.Labels),
+		})
+	}
+
+	return o.w.Write(ctx, querierpb.EvaluateQueryResponse{
+		Message: &querierpb.EvaluateQueryResponse_SeriesMetadata{
+			SeriesMetadata: &querierpb.EvaluateQueryResponseSeriesMetadata{
+				NodeIndex: o.nodeIndex,
+				Series:    protoSeries,
+			},
+		},
+	})
+}
+
+func (o *evaluationObserver) InstantVectorSeriesDataEvaluated(ctx context.Context, evaluator *streamingpromql.Evaluator, seriesIndex int, seriesData types.InstantVectorSeriesData) error {
+	defer types.PutInstantVectorSeriesData(seriesData, evaluator.MemoryConsumptionTracker)
+
+	// TODO: batch up series to return, rather than sending each series one at a time?
+
+	return o.w.Write(ctx, querierpb.EvaluateQueryResponse{
+		Message: &querierpb.EvaluateQueryResponse_InstantVectorSeriesData{
+			InstantVectorSeriesData: &querierpb.EvaluateQueryResponseInstantVectorSeriesData{
+				NodeIndex:  o.nodeIndex,
+				Floats:     mimirpb.FromFPointsToSamples(seriesData.Floats),
+				Histograms: mimirpb.FromHPointsToHistograms(seriesData.Histograms),
+			},
+		},
+	})
+}
+
+func (o *evaluationObserver) RangeVectorStepSamplesEvaluated(ctx context.Context, evaluator *streamingpromql.Evaluator, seriesIndex int, stepIndex int, stepData *types.RangeVectorStepData) error {
+	// We do not need to return anything to the pool here: the ring buffers in stepData are reused for subsequent steps, or returned to the pool elsewhere if not needed.
+
+	// TODO: batch up series / steps to return, rather than sending each step one at a time?
+
+	floatsHead, floatsTail := stepData.Floats.UnsafePoints()
+	floats, needToReturnFloats, err := combineSlices(floatsHead, floatsTail, types.FPointSlicePool, evaluator.MemoryConsumptionTracker)
+	if err != nil {
+		return err
+	}
+
+	if needToReturnFloats {
```

Comment:
I was suggesting to use cleanup func(){} approach instead of return flag like this. The suggestion applies also for histogram pool below.

Reply:
Good idea, changed in fa01766.